### PR TITLE
Python async pub example

### PR DESCRIPTION
**Pull request by xFrenchy**
### Summary
Python example displaying async publication

### Details and comments
!important! This example is not ready yet, creating initial pull request for review purposes. 
Here is a list of things to review:
-Idl file and class cannot be called "async" for python since it's a keyword. Naming the struct async and running rtiddsgen won't cause an error, however you won't be able to "from async import async". I have renamed it to async_type. Let me know if this should be approached differently
-When checking the history depth with Admin Console for the DataWriter, it shows a depth of 1 when it should be 12. This is observable when setting the QoS via XML and/or setting it programatically
-When creating a flow controller, I'm not sure what FlowController::FIXED_RATE_NAME is and what the python equivalent would be, for now I've created a flow controller without specifying a name

### Checks

-   [ ] I have updated the documentation accordingly.
-   [x] I have read the [CONTRIBUTING](https://github.com/rticommunity/rticonnextdds-examples/blob/develop/CONTRIBUTING.md) document.

### CODE REVIEW

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        try:
+            # Modify the data to be sent here
+            sample.x = count
+            print("Writing async_type, count {}".format(count))
+            writer.write(sample)
+            time.sleep(1)
+        except KeyboardInterrupt:
+            break
+    print("preparing to shut down...")
+    writer.close()
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:70]: explicit calls to close are not necessary

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    writer = dds.DataWriter(participant.implicit_publisher, topic, writer_qos)
+
+    # Create data sample for writing
+    sample = async_type(x=42)
+
+    for count in range(sample_count):
+        try:
+            # Modify the data to be sent here
+            sample.x = count
+            print("Writing async_type, count {}".format(count))
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:64]: use f-strings for formatting

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:1]: I think you should base the example on the C# one, which is more complete:

- it provides an option to use XML or programmatic Qos
- It uses WaitForAsyncPublication

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]

**Comment by alexcamposruiz** [comment]
You should now target master instead of develop, since this example will work with Connext 7.0.0

**Comment by lulivi** [comment]
Hi @xFrenchy ! Can you apply the `black` formatting to this example?

You can run:

```bash
pip install black==22.3.0
black -l 79 --preview <file or directory>
```

**Comment by xFrenchy** [comment]
I applied black formatting (I think). I left a few notes in the publisher and subscriber files of things that I've found that I think are problematic. Right now the most problematic is the subscriber throwing an exception when shutting down the application. Otherwise I think it follows the C# example better than the previous attempt. I added an additional argument to use the QoS file or not. I didn't see this implemented in the C# example though. I see the variable, but I never see an argument or a way to toggle that variable in the example.

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]
I'm leaving some feedback. I still have to complete the review and test or confirm the issues you mention throughout the code.

File: examples/connext_dds/asynchronous_publication/py/async_subscriber.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        print("async subscriber sleeping")
+        rti.asyncio.run(process_data(reader))
+        # await process_data(reader)
+    except KeyboardInterrupt:
+        pass
+
+    print("preparing to shut down...")
+    # rti.asyncio.close(process_data(reader))
+    # Note: Exception thrown here. I assume it's because the coroutine is still running
+    participant.close()
```
[the remark this answers is not in the document]
Comment (xFrenchy) [reply examples/connext_dds/asynchronous_publication/py/async_subscriber.py]: Yup, my bad. I don't know why I keep forgetting that. I removed both.

File: examples/connext_dds/asynchronous_publication/py/async_type.idl [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ * RTI grants Licensee a license to use, modify, compile, and create derivative
+ * works of the Software.  Licensee has the right to distribute object form
+ * only for use with RTI products.  The Software is provided "as is", with no
+ * warranty of any type, including any warranty for fitness for any purpose.
+ * RTI is under no obligation to maintain or support the Software.  RTI shall
+ * not be liable for any incidental or consequential damages arising out of the
+ * use or inability to use the software.
+ */
+
+ struct async_type {
```
[the remark this answers is not in the document]
Comment (xFrenchy) [reply examples/connext_dds/asynchronous_publication/py/async_type.idl:13]: Should I then remove the IDL from the repo?

File: examples/connext_dds/asynchronous_publication/py/async_type.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# WARNING: THIS FILE IS AUTO-GENERATED. DO NOT MODIFY.
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_type.py:1]: Replace this with the (c) header. It's actually OK to modify this file.

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    if use_xml_qos:
+        # Retrieve the default DataWriter QoS, from USER_QOS_PROFILES.xml
+        writer_qos = dds.QosProvider.default.datawriter_qos
+    else:
+        # Configure the DataWriterQos in code, to the same effect as the XML file.
+        writer_qos = participant.default_datawriter_qos
+        writer_qos.reliability = dds.Reliability.reliable(
+            dds.Duration.from_seconds(60)
+        )
+        # Note: Admin Console is reporting history depth of 1 still regardless of changes in the QoS file or programatically
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:37]: I verified that the depth is 12, but the propagated BuiltinTopicData says it's 1, even in a C++ app. 

Did you see any issue in the example related to incorrect History Qos? If not, you can delete this note.

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        writer_qos.history.kind = dds.HistoryKind.KEEP_LAST
+        writer_qos.history.depth = 12
+        writer_qos.data_writer_protocol.rtps_reliable_writer.min_send_window_size = (
+            50
+        )
+        writer_qos.data_writer_protocol.rtps_reliable_writer.max_send_window_size = (
+            50
+        )
+
+        # Note: Without specifying a name it defaults to DDS_DEFAULT_FLOW_CONTROLLER_NAME. Do I need to specify the name?
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:48]: Yes, you have to specify the name. We need to add a constant with the fixed rate name, but for now use this string.

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        3: dds.Verbosity.STATUS_ALL,
+    }
+
+    # Sets Connext verbosity to help debugging
+    verbosity = verbosity_levels.get(args.verbosity, dds.Verbosity.EXCEPTION)
+
+    dds.Logger.instance.verbosity = verbosity
+
+    try:
+        # Note: C# example doesn't seem to allow using the use_xml_qos option?
```
Comment (alexcamposruiz) [inline examples/connext_dds/asynchronous_publication/py/async_publisher.py:126]: It looks like this example didn't include that option. It's not a big deal.

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]

File: examples/connext_dds/asynchronous_publication/py/async_publisher.py (hunk 7 above)
Comment (xFrenchy) [reply examples/connext_dds/asynchronous_publication/py/async_publisher.py:37]: I didn't see any issue, it was more of an observation I found when trying to confirm that the QoS changes I was making were correctly applying.

**Review by alexcamposruiz — APPROVED** [review]

**Pull request merged**